src/wallet/hdwallet.ts

```diff
@@ -197,4 +210,70 @@ export abstract class AbstractHDWallet {
             throw new Error(`Invalid zone: ${zone}`);
         }
     }
+
+    // Returns the HD node that derives the address.
+    // If the address is not found in the wallet, an error is thrown.
+    protected _getHDNodeForAddress(addr: string): HDNodeWallet {
+        const addressInfo = this._addresses.get(addr);
+        if (!addressInfo) {
+            throw new Error(`Address ${addr} is not known to this wallet`);
+        }
+
+        // derive a HD node for the from address using the index
+        const accountNode = this._accounts.get(addressInfo.account);
+        if (!accountNode) {
+            throw new Error(`Account ${addressInfo.account} not found`);
+        }
+        const changeNode = accountNode.deriveChild(0);
+        return changeNode.deriveChild(addressInfo.index);
+    }
+
+    public async signMessage(address: string, message: string | Uint8Array): Promise<string> { 
+        const addrNode = this._getHDNodeForAddress(address);
+        return await addrNode.signMessage(message);
+    }
+
+    public async signMessageBytes(address: string, message: Uint8Array): Promise<string> {
+        return await this.signMessage(address, message);
+    }
+
+    public async serialize(): Promise<SerializedHDWallet> { 
+        const addresses = Array.from(this._addresses.values());
+        return {
+            version: (this.constructor as any)._version,
+            phrase: this._root.mnemonic!.phrase,
+            coinType: this.coinType(),
+            addresses: addresses,
+            provider: this.provider,
+        };
+    }
+
+    static async deserialize<T extends AbstractHDWallet>(
+        this: new (root: HDNodeWallet, provider?: Provider) => T,
+        serialized: SerializedHDWallet
+    ): Promise<QuaiHDWallet|QiHDWallet> {
+        if (serialized.version !== (this as any)._version) {
+            throw new Error(`Invalid version ${serialized.version} for wallet (expected ${(this as any)._version})`);
+        }
+        if (serialized.coinType !== (this as any)._coinType) {
+            throw new Error(`Invalid coinType ${serialized.coinType} for wallet (expected ${(this as any)._coinType})`);
+        }
+
+        const mnemonic = Mnemonic.fromPhrase(serialized.phrase);
+        const path = (this as any).parentPath(serialized.coinType);
+        const root = HDNodeWallet.fromMnemonic(mnemonic, path );
+        const wallet = new this(root, serialized.provider);
+        
+        for (const addressInfo of serialized.addresses) {
+            wallet._addresses.set(addressInfo.address, addressInfo);
+        }
+        
+        for (const address of serialized.addresses) {
+            if (!wallet._accounts.has(address.account)) {
+                wallet.addAccount(address.account);
+            }
+        }
```

Comment:
Would it be overkill to just call the `addAddress` method here? It does some extra work that may not be needed but it will also do some additional validation. I am thinking about the case where someone doesnt store all the address info, calling the `addAddress` method will ensure all address info is properly assigned